[mlir][vector] Update docs for scalable vectors (#101842)
llvm/llvm-project: mlir/docs/Dialects/Vector.md

### LLVM level

On CPU, the `n-D` `vector` type currently lowers to `!llvm<array<vector>>`.
More concretely,
* `vector<4x8x128xf32>` lowers to `!llvm<[4 x [ 8 x < 128
x float >]]>` (fixed-width vector), and
* `vector<4x8x[128]xf32>` lowers to `!llvm<[4 x [ 8 x < vscale x 128
x float >]]>` (scalable vector).

There are tradeoffs involved related to how one can access subvectors and how
one uses `llvm.extractelement`, `llvm.insertelement` and `llvm.shufflevector`.
The section on [LLVM Lowering Tradeoffs](#llvm-lowering-tradeoffs) offers a
deeper dive into the current design choices and tradeoffs.

> Note, while LLVM supports vectors of scalable vectors, these are required to be

Author: I did, thanks!

fixed-width arrays of 1-D scalable vectors. This means, effectively, that
scalable vectors with a non-trailing scalable dimension (e.g.
`vector<4x[8]x128xf32`) are not convertible to LLVM.

> Finally, as a brief reminder, MLIR takes similiar view on scalable Vectors as
> LLVM does (c.f. (Vector Type)[https://llvm.org/docs/LangRef.html#vector-type]):
> > The size of a specific scalable vector type is thus constant within IR, even
> > if the exact size in bytes cannot be determined until run time.
> 
> Specifically, the size of a scalable vector is not known at compile time, but
> known and fixed at run time

Reviewer: few suggestions
```suggestion
Finally, MLIR takes the same view on scalable Vectors as
LLVM (c.f. (Vector Type)[https://llvm.org/docs/LangRef.html#vector-type]):
> For scalable vectors, the total number of elements is a constant multiple
> (called vscale) of the specified number of elements; vscale is a positive
> integer that is unknown at compile time and the same hardware-dependent
> constant for all scalable vectors at run time.  The size of a specific
> scalable vector type is thus constant within IR, even if the exact size in
> bytes cannot be determined until run time.
```

Author: That paragraph from LLVM is great, but I'd still like to emphasise the key point, i.e. I'd like to keep this:

> Specifically, the size of a scalable vector is not known at compile time, but
known and fixed at run time.

Mostly to reinforce the message. I do feel that there's been a bit of confusion between "dynamic" and "scalable" vectors in MLIR. WDYT?

Reviewer: I think the LLVM one is bulletproof so I'd just copy that verbatim tbh. If your aim is to address confusion around dynamic vs scalable I dont think this does that anyway, but I did notice you updated "MLIR does not currently support dynamic vector sizes (i.e. SVE style)" elsewhere, perhaps that could be kept but changed to "fully-dynamic"?

Author: >  perhaps that could be kept but changed to "fully-dynamic"

Personally, I'd find that confusing unless we were to define "fully-dynamic". I'm deliberately trying to avoid that mindful that [[RFC] Dynamic Vector Semantics for the MLIR Vector Dialect](https://discourse.llvm.org/t/rfc-dynamic-vector-semantics-for-the-mlir-vector-dialect/75704) reached no consensus. Instead, I prefer to focus on defining what _is supported_.

Let me remove my note then.

### LLVM level

On CPU, the `n-D` `vector` type currently lowers to `!llvm<array<vector>>`.
More concretely,
* `vector<4x8x128xf32>` lowers to `!llvm<[4 x [ 8 x < 128
x float >]]>` (fixed-width vector), and
* `vector<4x8x[128]xf32>` lowers to `!llvm<[4 x [ 8 x < vscale x 128
x float >]]>` (scalable vector).

There are tradeoffs involved related to how one can access subvectors and how
one uses `llvm.extractelement`, `llvm.insertelement` and `llvm.shufflevector`.
The section on [LLVM Lowering Tradeoffs](#llvm-lowering-tradeoffs) offers a
deeper dive into the current design choices and tradeoffs.

Note, while LLVM supports arrarys of scalable vectors, these are required to be
fixed-width arrays of 1-D scalable vectors. This means scalable vectors with a
non-trailing scalable dimension (e.g. `vector<4x[8]x128xf32`) are not
convertible to LLVM.

> Finally, as a brief reminder, MLIR takes similiar view on scalable Vectors as
> LLVM does (c.f. (Vector Type)[https://llvm.org/docs/LangRef.html#vector-type]):

Reviewer: ```suggestion
Finally, MLIR takes the same view on scalable Vectors as
LLVM (c.f. (Vector Type)[https://llvm.org/docs/LangRef.html#vector-type]):
```

`vector<s_0x...s_{n-1}xf32>`). Lowering such an `n-D` MLIR vector type to an
LLVM descriptor can be done by either:

> 1.  Nested aggregate type of `1-D` vector:

Reviewer: is the order here is important?

Author: If you scroll further down, you will see that Option 1 and Option 2 are discussed in the reverse order. I just wanted to improve the flow (also got confused myself when reading this). 

I can move this change to a separate patch, but am concerned that that's just extra noise.
